crates/error/src/macros.rs

```diff
@@ -0,0 +1,342 @@
+//! Macro definitions and the private runtime functions used in their generated
+//! code.
+
+// Items used by macro-generated code.
+pub use core::format_args;
+pub use core::result::Result::Err;
+
+use super::{Error, OutOfMemory};
+use alloc::string::String;
+use core::fmt::{self, write};
+
+/// Construct an [`Error`](crate::Error) via string formatting or another error.
+///
+/// Like `anyhow::anyhow!` but for [`wasmtime::Error`][crate::Error].
+///
+/// # String Formatting
+///
+/// When a string literal is the first argument, it is interpreted as a format
+/// string template and the rest of the arguments are format arguments:
+///
+/// ```
+/// # use wasmtime_internal_error as wasmtime;
+/// use wasmtime::{anyhow, Error};
+///
+/// let x = 42;
+/// let error: Error = anyhow!("x is {x}");
+/// assert_eq!(error.to_string(), "x is 42");
+///
+/// let error: Error = anyhow!("x / 2 is {}", x / 2);
+/// assert_eq!(error.to_string(), "x / 2 is 21");
+///
+/// let error: Error = anyhow!("x + 1 is {y}", y = x + 1);
+/// assert_eq!(error.to_string(), "x + 1 is 43");
+/// ```
+///
+/// # From Another Error
+///
+/// When a string literal is not the first argument, then it is treated as a
+/// foreign error and is converted into an [`Error`][crate::Error]. The argument
+/// must be of a type that can be passed to either
+/// [`Error::new`][crate::Error::new] or [`Error::msg`][crate::Error::msg]:
+///
+/// ```
+/// # fn _foo() {
+/// #![cfg(feature = "std")]
+/// # use wasmtime_internal_error as wasmtime;
+/// use std::fmt;
+/// use wasmtime::{anyhow, Error};
+///
+/// #[derive(Debug)]
+/// struct SomeOtherError(u32);
+///
+/// impl fmt::Display for SomeOtherError {
+///     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+///         write!(f, "some other error (code {})", self.0)
+///     }
+/// }
+///
+/// impl std::error::Error for SomeOtherError {}
+///
+/// let error: Error = anyhow!(SomeOtherError(36));
+/// assert!(error.is::<SomeOtherError>());
+/// assert_eq!(error.to_string(), "some other error (code 36)");
+/// # }
+/// ```
+#[macro_export]
+macro_rules! anyhow {
```

Comment:
Could we actually delete this macro and use `format_err` instead? I'm otherwise a bit worried about copying `anyhow` too too closely here and causing confusion. Basically the name `anyhow` feels like it should be reserved for `anyhow`-the-crate, while the other macros here are all generic enough it seems fine to use them